[AOTInductor] Add Runtime Constant-folding for AOTInductor

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113243

Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

muchulee8 committed:
[TEST]

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[TEST]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
TBD

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52430653).

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1939,13 +2065,16 @@ def use_thread_local_cached_output_tensor(idx, output):
+                result.writeline("} // AOTInductorModel::_const_run_impl")
+            else:
+                result.writeline("} // namespace aot_inductor")
+                result.writeline("} // namespace torch")
```
Why do we not place AOTInductorModel::_const_run_impl in the ```torch``` namespace?

muchulee8 replied:
It is in the `torch` namespace. The whole logic is, if `is_const_graph`, we would generate a constant folding code. The `generate_end` for the constant folding code is the closing of that part.
If it's the main codegen, constant folding code is part of the "main", the `generate_end` for `else` is closing the `main`.

chenyang78 commented on torch/_inductor/ir.py:
```diff
@@ -4326,9 +4326,13 @@ def __init__(self, x, indices, values, accumulate):
+                (r.name in V.graph.constants and isinstance(r, dependencies.MemoryDep))
+                for r in x.get_reads()
+            )
+            and (not V.graph.aot_mode or not config.use_runtime_constant_folding)
```
Why do we include ```not V.graph.aot_mode```?

muchulee8 replied:
Changed.

chenyang78 commented on torch/fx/passes/split_utils.py:
```diff
@@ -274,7 +278,7 @@ def remap_func(x):
         )
 
-        if len(outs) == 1:
+        if len(outs) == 1 and not return_tuple:
```
Hmm, wondering why we need this extra ```not return_tuple``` condition.

muchulee8 replied:
https://github.com/pytorch/pytorch/blob/main/torch/fx/graph.py#L1264

We have an implicit assumption that the return arg should be a single tuple.

chenyang78 reviewed (commented):
Left a number of comments. I don't feel confident enough to review the changes from constant_folding.py. @eellison would be a better person to take a look. Thanks!

eellison reviewed (commented):
would you mind expanding the pr description ?

desertfire commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +200,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+        if node.op == "get_attr":
+            used_to_fold = False
+            for u in node.users:
+                if u.tag == CONST_MODULE_TAG:
```
I think this is correct because `CONST_MODULE_TAG` has been properly populated by `ConstantFolder`. Is there a good to test it? Maybe adding some logging and verify when testing?

desertfire commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +200,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+
+        for user in node.users:
+            if user.tag == MODULE_TAG:
+                output_nodes.append(new_node)
```
Why only append when `node` is used by a non-folded node?

muchulee8 replied:
Those "outputs" are the nodes that are "constants being folded", which would then be consumed by the main graph. Is your concern that we're adding too many or are we missing nodes?

khabinov commented on torch/_inductor/codecache.py:
```diff
@@ -1637,7 +1641,9 @@ def _to_bytes(t: torch.Tensor) -> bytes:
-                        _to_bytes(tensor) for tensor in graph.constants.values()
+                        _to_bytes(tensor)
+                        for name, tensor in graph.constants.items()
+                        if name not in graph.folded_constants
```
> Do we expect to have different constant names due to realtime-weight update?

Constant names should remain the same.

> With realtime-weight update, will any weight stay constant or should we treat all weight may change?

We should expect that all weights will change as the model gets retrained.

khabinov commented on torch/_inductor/codegen/aoti_runtime/interface.cpp:
```diff
@@ -125,7 +125,23 @@ AOTIRuntimeError AOTInductorModelContainerUpdateInactiveConstantBuffer(
           constant_map_handle,
           /*use_inactive*/ true,
-          /*validate_full_update*/ true);
+          /*validate_full_update*/ false);
```
Yeah, it would still be great to validate that all weights were provided even if some of them will be folded eventually.

muchulee8 replied:
Done.

khabinov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1516,64 +1522,93 @@ def write_wrapper_decl(self):
+                self.header.splice(V.graph.const_graph.wrapper_code.header)
+                self.prefix.splice(V.graph.const_code)
+
+            if V.graph.is_const_graph:
```
Agree, naming might be a bit confusing.

muchulee8 replied:
Changed the name from `constant_graph` to `const_module`. LMK if it sounds better.
`is_constant_graph` means that the lowering module for this class is a constant graph (without inputs.)
`const_graph`/`const_module` is a attribute in this class. It's used to embed the code into the to-be-generate code that this class will be generating.

khabinov reviewed (commented):
Wondering what commit this PR is based on? Would it be possible to rebase?

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1582,19 +1617,20 @@ def write_wrapper_decl(self):
-                                auto inputs = alloc_tensors_by_stealing_from_handles(input_handles, num_inputs());
-                            """
-                        )
+                    if not V.graph.is_const_graph:
```
Not really, since merging them might cause us accidentally branching into the `else` branch below.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1516,64 +1522,93 @@ def write_wrapper_decl(self):
+                            std::vector<AtenTensorHandle>& output_handles,
+                            DeviceStreamType stream,
+                            AOTIProxyExecutorHandle proxy_executor
+                        ) {}
```
IMO it might not be that trivial or cleaner at this moment (we still need to do branching, just at some other places.) I feel like we should do a full clean up for ALL branches within wrapper holistically afterwards.

chenyang78 replied:
OK, sounds good. I will be working on a holistic refactoring soon.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1658,11 +1696,12 @@ def write_wrapper_decl(self):
-                    self.codegen_input_numel_asserts()
-                else:
-                    self.prefix.writeline("inputs.clear();")
+                if not V.graph.is_const_graph:
```
Not necessarily. The point here is that, for const graph's codegen, we are dropping all input related codegen.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1702,9 +1741,10 @@ def codegen_model_kernels(self):
-        for kernel in chain(
-            self.src_to_kernel.values(), self.user_defined_kernel_cache.values()
-        ):
+        declare_kernel = set(self.src_to_kernel.values())
```
I don't think it matters w.r.t. `chain`, since the values are just the keys' string (which would be plugged in `set`). I use set here so that we won't "redeclare" kernels that are shared by both constant's code and main code.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1800,10 +1844,90 @@ def escape_string(x):
+            _const_run_impl(output_handles, stream, proxy_executor);
+            // build up folded_constants_map
+            return folded_constants_map;
+        }
```
This is not a boilerplate code. It's a comment that shows what code might be generated with this function.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1939,13 +2065,16 @@ def use_thread_local_cached_output_tensor(idx, output):
 
     def generate_before_suffix(self, result):
-        if V.graph.aot_mode:
+        if V.graph.aot_mode and not V.graph.is_const_graph:
```
I think this condition is only used twice?

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -384,8 +384,9 @@ def __init__(self):
 
-        self.write_header()
-        self.write_prefix()
+        if not V.graph.is_const_graph or not V.graph.cpp_wrapper:
```
Moved into subclass.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1800,10 +1844,90 @@ def escape_string(x):
+            ) {
+            """
+        )
+        if not config.use_runtime_constant_folding:
```
I would propose we do the refactor all-in-once that deals with ALL branching logic in the codegen.

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1800,10 +1844,90 @@ def escape_string(x):
+                if name not in V.graph.const_output_index:
+                    continue
+                else:
+                    const_index_mapping[V.graph.const_output_index[name]] = (idx, name)  # type: ignore[call-overload]
```
Done.

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

chenyang78 reviewed (commented):
Could you write some high level description somewhere (e.g. interface.h?) that would help people understand:

* the purpose of the runtime constant-folding;
* the way of how it works

I think it would also help @desertfire and @eellison review the changes. Thanks!

muchulee8 commented on test/cpp/aot_inductor/test.py:
```diff
@@ -5,39 +5,51 @@
-    })
+    for use_runtime_constant_folding in [True, False]:
+        if device == "cpu" and use_runtime_constant_folding:
+            # We do not test runtime const folding for cpu mode.
```
We can. The piece we are missing at the moment is the way we store constants (we now directly point our weight to the constant we embed in the .so, we would want to separate that to use different buffer.) It would be more direct and easier if we later split out the constants from .so.

muchulee8 commented on torch/_inductor/codecache.py:
```diff
@@ -1544,6 +1544,10 @@ def set(cls, key: str, params: Dict[str, str], cubin: str) -> None:
         return cls.cache.get(key, None)
 
+    @classmethod
+    def get_keys(cls):
```
Yes, it's not used "functionally", I added this to improve our error message in the `generate_default_grid` in wrapper.py.

muchulee8 commented:
> Could you write some high level description somewhere (e.g. interface.h?) that would help people understand:
> 
> * the purpose of the runtime constant-folding;
> * the way of how it works
> 
> I think it would also help @desertfire and @eellison review the changes. Thanks!

I added the purpose in the comment of `constant_folding.py`. And the updated summary has the implementation details. Thanks!

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1424,6 +1424,10 @@ def write_constant(self, name, hashed):
         self.header.writeline(f"// {name} {hashed}")
 
     def write_header(self):
+        if V.graph.is_const_graph:
+            # We do not write header for constant graph, it will be written by main module.
```
Can we pass in `is_const_graph` to `WrapperCodeGen` ?

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +203,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+            node.tag = MODULE_TAG
+
+
+def get_constant_graph(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
```
nit: `run_and_get_constant_graph` maybe ? or something else that will show that we are running constant folding.

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +203,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+                output_nodes.append(new_node)
+                break
+
+    new_graph.output(tuple(output_nodes))
```
I would lint the new graph

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +203,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+    # any folding opportunity, we keep it in main gm.
+    for node in gm.graph.nodes:
+        if node.op == "get_attr":
+            used_to_fold = False
+            for u in node.users:
+                if u.tag == CONST_MODULE_TAG:
+                    used_to_fold = True
+                    break
+            if not used_to_fold:
+                node.tag = MODULE_TAG
```
Can't we remove the `node.op == "get_attr"` condition above in `constant_graph_tag` and remove this loop?

muchulee8 replied:
I don't think so, the "users" will only be tagged when all inputs are being tagged. So if this is happening at tagging time, right at this moment, the users aren't determined whether it's in const or main yet. Or are you asking to move this loop to `constant_graph_tag`?

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -1637,7 +1641,9 @@ def _to_bytes(t: torch.Tensor) -> bytes:
 
             aot_constants = b"".join(
-                _to_bytes(tensor) for tensor in graph.constants.values()
+                _to_bytes(tensor)
+                for name, tensor in graph.constants.items()
+                if name not in graph.folded_constants
```
Can you say more abou this ?

muchulee8 replied:
When we are compiling the constants into the .so, we have 2 kinds of constants. 
Recall we have 2 steps, 1) split out the constant graph 2) rewire the main graph
In step 2), to successfully to have the compilation succeed, we actually have constants attached on the main graph (which, are the constants resulting from step 1), the output of `constant_graph.run()`.) The constants being attached are the "folded constants", which we do not want to include in the final .so, since they can be determined during "runtime" (which is the purpose of this change.)

muchulee8 replied:
We only have one single .so being generated. So the .so file includes all the constants. The aot_constants here is constants that are being written into .so, so it doesn't matter whether the constant belongs to const graph or main graph, they are all in the single file, it's just the part of the code that want to use the constant need to grab it from the corresponding offset.

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -195,3 +203,60 @@ def constant_fold(gm, constraint_fn: Optional[Callable[[torch.fx.Node], bool]] =
+            or node in cf.node_replacements
+            or node in cf.replaced_uses
+        ):
+            node.tag = CONST_MODULE_TAG
+        else:
+            node.tag = MODULE_TAG
+
```
we typically put important metadata of nodes in `node.meta`. or we could maintain dictionaries ourself.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -578,16 +599,17 @@ def mark_buffer_mutated(self, name: str):
-                    and torch.eq(data, value).all()
-                ):
-                    return constant_name
+            if not config.aot_inductor.use_runtime_constant_folding:
```
Why dont we use this instead: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/config.py#L282-L284

muchulee8 replied:
Ah.. I think it's slightly different. `always_keep_tensor_constants` prevents us inlining constants directly, however it allows reusing of the constants (this if statement checks if the to-be-added tensor has a value exactly the same as existing constants.)
We add the if here mainly because we want to make sure that all constants get generated "even if we have multiple constants with same value". Since we can't guarantee that constants wouldn't be updated.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -717,7 +740,11 @@ def get_attr(self, target, args, kwargs):
         value = getattr(self.module, target)
 
-        if config.always_keep_tensor_constants or unsupported_output_tensor(value):
+        if (
+            config.aot_inductor.use_runtime_constant_folding
+            or config.always_keep_tensor_constants
+            or unsupported_output_tensor(value)
+        ):
```
similarly, this should be unnecessary

eellison commented on torch/_inductor/graph.py:
```diff
@@ -211,11 +218,25 @@ def __init__(
         self.cuda = False
         self.buffers: List[ir.Buffer] = []
-        self.constants: Dict[str, torch.Tensor] = {}
+        self.const_output_index: Dict[str, int] = (
+            const_output_index if const_output_index else {}
+        )
+        self.const_kernels: Set[str] = (
+            const_module.wrapper_code.src_to_kernel.values() if const_module else set()
+        )
+        self.folded_constants: Set[str] = (
+            set(const_output_index.keys()) if const_output_index else set()
+        )
+        self.constants: Dict[str, torch.Tensor] = (
+            const_module.constants if const_module else {}
+        )
+        self.used_constants: Set[str] = set()
```
We're adding five new fields here.. can we think of other ways of doing this ?

eellison replied:
`used_constants` is any constant with a use. we can infer this from the graph instead of adding this field

eellison commented on torch/_inductor/graph.py:
```diff
@@ -211,11 +218,25 @@ def __init__(
         self.cuda = False
         self.buffers: List[ir.Buffer] = []
-        self.constants: Dict[str, torch.Tensor] = {}
+        self.const_output_index: Dict[str, int] = (
```
Can't we infer this from the graph ? We know from the graph which output name corresponds to which name.

muchulee8 replied:
I would say yes now. We can definitely match the output name that we have prefix of `_FOLDED_CONST_` and match with the node name. I've chose to err on the safe side to just pass information from upstream which constant graph is generated, but I'm also fine to hardcode the string if you prefer.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -211,11 +218,25 @@ def __init__(
+        self.const_output_index: Dict[str, int] = (
+            const_output_index if const_output_index else {}
+        )
+        self.const_kernels: Set[str] = (
+            const_module.wrapper_code.src_to_kernel.values() if const_module else set()
```
This is never used.

muchulee8 replied:
This is used in `wrapper.py`. This is used in the case that we might have duplicated kernels between constant module and main module.

eellison reviewed (commented):
Left a number of comments. Two main concerns:

1) It's surprising that we are making inductor lowering aware of so many details of constant folding. Maybe we can drill down on specific examples, but it's not something that I would expect is needed, especially with `always_keep_tensor_constants` already as an option.

2) I would much prefer we did not `config.constant_folding` as an option which duplicates `freezing` which already exists. When you add `freezing` to max autotune it adds 14% to TB, 10% to HF, 21% to TIMM. While it has a few more features and is a little more complicated than constant folding, that's how these things start. Next thing constant folding is going to want to handle converting convolution weights to channels last and it's going to start to replicate more of freezing. I would like to brainstorm what it would take to not duplicate codepaths/config options.

muchulee8 commented on torch/_inductor/compile_fx.py:
```diff
@@ -170,6 +170,62 @@ def _unlift_graph(mod, gm, graph_signature):
+    const_gm = get_constant_graph(gm)
+    const_result = const_gm()
+
+    const_names = {
```
How? If you look into line 204/line 214/etc. all of those are directly using names.

muchulee8 commented on torch/_inductor/constant_folding.py:
```diff
@@ -6,22 +6,30 @@
 
+# We would like to split modules into two subgraphs for runtime weight updates to work correctly.
+# The use case and more information could be found at:
+# https://docs.google.com/document/d/1inZC-8KarJ6gKB7G9egmYLx1V_dKX_apxon0w4zPC0Q/edit?usp=sharing
```
I'm also not 100% happy with adding internal docs, but as @desertfire have requested, adding more information also makes sense to me. I did a grep and do see that this practice exists within Pytorch repo.

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_runner/model_container_runner.h:
```diff
@@ -35,6 +35,9 @@ class TORCH_API AOTIModelContainerRunner {
       bool validate_full_updates);
+  void run_const_fold(
+      bool use_inactive,
+      AOTInductorStreamHandle cuda_stream_handle = nullptr);
```
You can add pybind for this in https://github.com/pytorch/pytorch/blob/main/torch/csrc/inductor/aoti_runner/pybind.cpp, and then you should be able to test in python.

desertfire reviewed (commented):
Still feel intimidating to review such a big PR. Would be helpful to see some example before/after generated code.

muchulee8 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -187,6 +187,23 @@ def forward(self, x, y):
+                return torch.matmul(x, w)
+
+        example_inputs = (torch.randn(4, 4, device=self.device),)
+        with config.patch({"aot_inductor.use_runtime_constant_folding": True}):
```
I guess we would need to add the pybind, but I would say we should keep that away too since it's not directly related to this PR. (And as you mentioned it's a big PR already...)

muchulee8 commented:
> Still feel intimidating to review such a big PR. Would be helpful to see some example before/after generated code.

P1123973060 This is the diff for the python test added (one difference is that I changed the "bias" into a 4x4, otherwise it would be directly expanded.) 
Basically it's just adding corresponding code into the run const folding part.

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1848,10 +1895,88 @@ def escape_string(x):
+        if not config.aot_inductor.use_runtime_constant_folding:
+            self.prefix.splice(
+                """
+                    return {};
```
It's better to raise a runtime error here. If a user calls `runner.run_const_fold` on a model.so compiled with `use_runtime_constant_folding = False`, it should give a loud error instead of silently pass.

desertfire commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -80,6 +81,26 @@ class AOTInductorModelContainer {
     std::shared_lock model_lk(model_exec_mutex_);
     auto* model = get_available_model();
+
+    if (!constant_folded_) {
```
Why not doing this in the initializer?

desertfire reviewed (approved):
Please address my new comments before merging.

Also I think the ABI-compatible mode is not tested here. Since it's a default off feature, I am ok with adding more tests in the following PRs.

chenyang78 reviewed (approved):
LGTM. Thanks!

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
Update on "[AOTInductor] Add Runtime Constant-folding for AOTInductor"


Summary:
Add Runtime Constant-folding for AOTInductor.
This also include the invocation of constant folding at load time.

The constant folding lowering is a 2-step process. 
First, we split the graph into 2 modules, one of it is the constant module, which doesn't depend on any input and the whole module could be inferred (constant-folded) one-time and be reused. The constant module, is lowered, and being codegen-ed as usual and cached (let's call this constant code). The constant code reuses the whole lowering/profiling/etc. process, only difference is that we do not generate any headers or initialization for the constant code.
Second, after handling the constant module, we take care of the main module (which is the part that would depend on the user input.) For the main module, we take in one additional component, the constant code, compare with a normal lowering. Addition step we do here is that, we inject the constant code into the codegen-ed main module, and create the caller for the main module to consume the result of the constant module.

Test Plan:
Included in commit.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D52430653](https://our.internmc.facebook.com/intern/diff/D52430653)

[ghstack-poisoned]

eellison commented:
clearing queue, re-request review if you need it !
